File: FStarFSharp/CPEG/ParsingSemantics.fs
module ParsingSemantics

open Expressions

(*
type Tree =
    | TNode of l : string v : Tree
    | TConcat of vlist : Tree list
    | TEmpty
    | Fail
 *)

type Tree =
    | TNode of l : string * v : Tree
    | TConcat of vlist : Tree list
    | TSeq of vlist : Tree list
    | TEmpty
    | Fail
    | TLeaf of l : string * str : string


let rec parsing (se:CPEGExp) (sx:string) =
  match se with
    | Empty -> (TEmpty, sx)
    | Terminal a -> match sx with
                    | "" -> (Fail, "")
                    | _ -> if String.length sx <> 0 && a = sx.[0..(a.Length-1)] then (TEmpty, sx.[a.Length..sx.Length-1]) else (Fail, sx)
    | Nonterm e -> parsing e sx
    | Seq exs -> List.fold (fun v exp -> printfn "%A" v; match v with
                                | (Fail, ys) -> (Fail, ys)
                                | (TSeq oldsubt, xs) -> match parsing exp xs with
                                                | (Fail, _) -> (Fail, sx)
                                                | (TEmpty, ys) -> (TSeq oldsubt, ys)
                                                | (subt, ys) -> (TSeq (subt::oldsubt), ys)//subtをoldsubtの末尾に付くように直す
                                | (_, ys) -> (Fail, ys)
                           ) (TSeq [], sx) exs
                    |> fun (result, ys) -> match result with
                              | TSeq [] -> (TEmpty, ys)
                              | TSeq [x] -> (x, ys) 
                              | TSeq subt -> (TSeq (List.rev <| subt),ys)
                              | _ -> (result, ys)
    | Alt(e1,e2) -> match parsing e1 sx with
                    | (Fail,_) -> parsing e2 sx
                    | a -> a
    | Not e -> match parsing e sx with
                    | (Fail, _) -> (TEmpty, sx)
                    | _ -> (Fail, sx)
    | Rep e -> let mutable inputstr = sx
               let mutable treelist = List.Empty
               let mutable loopcondition = true
               while loopcondition do
                    let newtree, newstr = parsing e inputstr
                    //printfn "%A" newstr
                    inputstr <- newstr
                    if newtree = Fail then 
                      loopcondition <- false 
                    if (newtree <> TEmpty && newtree <> Fail) then
                      treelist <- newtree :: treelist
               match treelist with
                | [] -> (TEmpty, inputstr)
                | [x] -> (x, inputstr)
                | _ -> (TConcat (List.rev treelist), inputstr)
    | Cap(l, e) -> match parsing e sx with
                    | (Fail, _) -> (Fail, sx)
                    | (v, ys) -> (TNode(l, v), ys)
    | FCap(e1, l, e2) -> match parsing e1 sx with
                            | (Fail, _) -> (Fail, sx)
                            | (v, ys) -> match parsing (Seq [e2; Rep e2]) ys with
                                    | (TSeq w, zs) -> let foldTree = List.fold (fun lTree rTree -> TNode(l,TConcat [lTree; rTree])) v w in (foldTree, zs)
                                    | _ -> (v, ys)
    | Leaf(l, e) -> let lenE = e.Length
                    //printfn "call-, %A" lenE
                    if String.length sx <> 0 && e = sx.[0..lenE-1] then (TLeaf(l,e), sx.[lenE..sx.Length-1]) else (Fail, sx)

File: FStarFSharp/CPEG/ParsingSemantics.py
from Expressions import *
import functools

#type Tree =
#    | TNode of l : string v : Tree
#    | TConcat of vlist : Tree list
#    | TEmpty
#    | Fail

class Tree:
    def __init__(self):
        pass

class TNode(Tree):
    def __init__(self, v1, v2):
        self.l = v1
        self.v = v2

class TConcat(Tree):
    def __init__(self, v):
        self.vlist = v

class TSeq(Tree):
    def __init__(self, v):
        self.vlist = v

class TEmpty(Tree):
    def __init__(self):
        pass

class Fail(Tree):
    def __init__(self):
        pass

class TLeaf(Tree):
    def __init__(self, v1, v2):
        self.l = v1
        self.str = v2


def parsing (se: CPEGExp, sx: str):
    if isinstance(se, Empty):
        return (TEmpty(), sx)
    elif isinstance(se, Terminal):
        if len(sx) == 0:
            return (Fail(), "")
        else:
            return (TEmpty(), sx[len(se.value):]) if len(sx) != 0 and a == sx[:len(se.value)] else (Fail(), sx)
    elif isinstance(se, Nonterm):
        return parsing(se.value, sx)
    elif isinstance(se, Seq):
        def closure(v, exp):
            if isinstance(v[0], Fail):
                return v
            elif isinstance(v[0], TSeq):
                t = parsing(exp, v[1])
                if isinstance(t[0], Fail):
                    return (Fail(), sx)
                elif isinstance(t[0], TEmpty):
                    return (v[0], t[1])
                else:
                    return ((TSeq([t[0]] + v[0].vlist)), t[1])#subtをoldsubtの末尾に付くように直す
            else:
                return (Fail(), v[1])

        (result, ys) = functools.reduce(closure, (TSeq([]), sx), se.elist)
        if isinstance(result, TSeq):
            if len(result.vlist) == 0:
                return (TEmpty(), ys)
            elif len(result.vlist) == 1:
                return (result.vlist[0], ys)
            else:
                return (TSeq(reversed(result.vlist)), ys)
        else:
            return (result, ys)
    elif isinstance(se, Alt):
        t = parsing(se.e1, sx)
        if isinstance(t[0], Fail):
            return parsing(se.e2, sx)
        else:
            return t
    elif isinstance(se, Not):
        t = parsing(se.e, sx)
        if isinstance(t[0], Fail):
            return (TEmpty(), sx)
        else:
            return (Fail(), sx)
    elif isinstance(se, Rep):
        inputstr = sx
        treelist = []
        loopcondition = True
        while loopcondition:
            (newtree, newstr) = parsing(se.e, inputstr)
            #printfn "%A" newstr
            inputstr = newstr
            if isinstance(newtree, Fail):
                loopcondition = False
            if (not isinstance(newtree, TEmpty)) and (not isinstance(newtree, Fail)):
                treelist.append(newtree)
        if len(treelist) == 0:
            return (TEmpty, inputstr)
        elif len(treelist) == 1:
            return (treelist[0], inputstr)
        else:
            return (TConcat(treelist), inputstr)
    elif isinstance(se, Cap):
        (v, ys) = parsing(se.e, sx)
        if isinstance(v, Fail):
            return (Fail(), sx)
        else:
            return (TNode(se.l, v), ys)
    elif isinstance(se, FCap):
        (v, ys) = parsing(se.e1, sx)
        if isinstance(v, Fail):
            return (Fail(), sx)
        else:
            (w, zs) = parsing(Seq([se.e2, Rep(se.e2)]), ys)
            if isinstance(w, TSeq):
                foldTree = reduce(lambda lTree, rTree: TNode(se.l, TConcat([lTree, rTree])), v, w.vlist)
                return (foldTree, zs)
            else:
                return (v, ys)
    elif isinstance(se, Leaf):
        lenE = len(se.e)
        #printfn "call-, %A" lenE
        return (TLeaf(se.l, se.e), sx[lenE:]) if len(sx) != 0 and se.e == sx[:lenE] else (Fail(), sx)
